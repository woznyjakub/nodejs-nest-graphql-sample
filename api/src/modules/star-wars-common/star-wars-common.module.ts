import { Module } from '@nestjs/common';

import { StarWarsCommonService } from './star-wars-common.service';

@Module({
  providers: [StarWarsCommonService],
  exports: [StarWarsCommonService],
})
export class StarWarsCommonModule {}
