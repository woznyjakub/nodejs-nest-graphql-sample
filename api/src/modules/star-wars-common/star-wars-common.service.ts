import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';

export type PagedResponse<Item> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Item[];
};

@Injectable()
export class StarWarsCommonService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  async fetchWithCaching<T extends object>(
    cacheKey: string,
    getterFn: () => Promise<T>,
  ): Promise<T> {
    const cachedData = await this.cacheManager.get<T>(cacheKey);
    if (cachedData) {
      return cachedData;
    }

    const data = await getterFn();
    await this.cacheManager.set(cacheKey, data);

    return data;
  }

  async getAllPages<T extends object>(
    cacheHandlerFn: (page: number) => string,
    getterFn: (page: number) => Promise<PagedResponse<T>>,
  ): Promise<T[]> {
    const limit = 25;
    let page = 1;

    const results: T[] = [];

    do {
      const data = await this.fetchWithCaching(cacheHandlerFn(page), () => getterFn(page));

      if (Array.isArray(data.results) || !results?.length) {
        results.push(...data.results);
      } else {
        break;
      }

      if (!data.next) {
        break;
      }
      page++;
    } while (page < limit);

    return results;
  }

  private buildCacheKey(
    resourceName: string,
    dataType: 'page' | 'id',
    value: string | number,
  ): string {
    return `swapi-${resourceName}-${dataType}-${value}`;
  }

  getVehicleCacheKey(id: number): string {
    return this.buildCacheKey('vehicle', 'id', id);
  }

  getVehiclesCacheKey(page: number): string {
    return this.buildCacheKey('vehicles', 'page', page);
  }

  getFilmCacheKey(id: number): string {
    return this.buildCacheKey('film', 'id', id);
  }

  getFilmsCacheKey(page: number): string {
    return this.buildCacheKey('films', 'page', page);
  }

  getStarshipCacheKey(id: number): string {
    return this.buildCacheKey('starship', 'id', id);
  }

  getStarshipsCacheKey(page: number): string {
    return this.buildCacheKey('starships', 'page', page);
  }

  getPlanetCacheKey(id: number): string {
    return this.buildCacheKey('planet', 'id', id);
  }

  getPlanetsCacheKey(page: number): string {
    return this.buildCacheKey('planets', 'page', page);
  }

  getSpeciesCacheKey(id: number): string {
    return this.buildCacheKey('species', 'id', id);
  }

  getSpeciesManyCacheKey(page: number): string {
    return this.buildCacheKey('speciesMany', 'page', page);
  }

  getPeopleCacheKey(page: number): string {
    return this.buildCacheKey('people', 'page', page);
  }
}
