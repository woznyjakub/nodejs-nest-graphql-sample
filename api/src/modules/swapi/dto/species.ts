import { Expose, Type } from 'class-transformer';
import { IsArray, IsNumber, IsString, ValidateIf, ValidateNested } from 'class-validator';

export class SpeciesResponse {
  @Expose()
  @IsString()
  name: string;

  @Expose()
  @IsString()
  classification: string;

  @Expose()
  @IsString()
  designation: string;

  @Expose()
  @IsString()
  average_height: string;

  @Expose()
  @IsString()
  skin_colors: string;

  @Expose()
  @IsString()
  hair_colors: string;

  @Expose()
  @IsString()
  eye_colors: string;

  @Expose()
  @IsString()
  average_lifespan: string;

  @Expose()
  @ValidateIf((_, v) => v !== null)
  @IsString()
  homeworld: string | null;

  @Expose()
  @IsString()
  language: string;

  @Expose()
  @IsArray()
  @IsString({ each: true })
  people: string[];

  @Expose()
  @IsArray()
  @IsString({ each: true })
  films: string[];

  @Expose()
  @IsString()
  created: string;

  @Expose()
  @IsString()
  edited: string;

  @Expose()
  @IsString()
  url: string;
}

export class SpeciesManyResponse {
  @Expose()
  @IsNumber()
  count: number;

  @Expose()
  @ValidateIf((_, v) => v !== null)
  @IsString()
  next: string | null;

  @Expose()
  @ValidateIf((_, v) => v !== null)
  @IsString()
  previous: string | null;

  @Expose()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => SpeciesResponse)
  results: SpeciesResponse[];
}
