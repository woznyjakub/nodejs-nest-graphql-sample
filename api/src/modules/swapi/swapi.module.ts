import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';

import { SwapiService } from './swapi.service';

@Module({
  imports: [HttpModule],
  providers: [SwapiService],
  exports: [SwapiService],
})
export class SwapiModule {}
