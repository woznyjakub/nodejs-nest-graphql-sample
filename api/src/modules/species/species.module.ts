import { Module } from '@nestjs/common';

import { StarWarsCommonModule } from '../star-wars-common/star-wars-common.module';
import { SwapiModule } from '../swapi/swapi.module';

import { SpeciesResolver } from './species.resolver';
import { SpeciesService } from './species.service';

@Module({
  imports: [SwapiModule, StarWarsCommonModule],
  providers: [SpeciesResolver, SpeciesService],
})
export class SpeciesModule {}
