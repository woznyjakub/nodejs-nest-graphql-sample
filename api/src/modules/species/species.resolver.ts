import { Resolver, Query, Args, Int } from '@nestjs/graphql';

import { SpeciesMany } from './entities/species-many.entity';
import { Species } from './entities/species.entity';
import { SpeciesService } from './species.service';

@Resolver(() => Species)
export class SpeciesResolver {
  constructor(private readonly vehiclesService: SpeciesService) {}

  @Query(() => SpeciesMany, { name: 'speciesMany' })
  findAll(@Args('page', { type: () => Int }) page: number): Promise<SpeciesMany> {
    return this.vehiclesService.findAll(page);
  }

  @Query(() => Species, { name: 'species' })
  findOne(@Args('id', { type: () => Int }) id: number): Promise<Species> {
    return this.vehiclesService.findOne(id);
  }
}
