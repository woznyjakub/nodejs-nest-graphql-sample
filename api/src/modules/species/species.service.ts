import { Injectable } from '@nestjs/common';

import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SpeciesResponse } from '../swapi/dto/species';
import { SwapiService } from '../swapi/swapi.service';

import { SpeciesMany } from './entities/species-many.entity';
import { Species } from './entities/species.entity';

@Injectable()
export class SpeciesService {
  constructor(
    private readonly swapiService: SwapiService,
    private readonly swCommonService: StarWarsCommonService,
  ) {}

  async findAll(page: number): Promise<SpeciesMany> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getSpeciesManyCacheKey(page),
      () => this.swapiService.getSpeciesMany(page),
    );

    return {
      count: swapiData.count,
      next: swapiData.next,
      previous: swapiData.previous,
      results: swapiData.results.map(this.mapSingleSpecies),
    };
  }

  async findOne(id: number): Promise<Species> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getSpeciesCacheKey(id),
      () => this.swapiService.getSpecies(id),
    );

    return this.mapSingleSpecies(swapiData);
  }

  private mapSingleSpecies(species: SpeciesResponse): Species {
    return {
      name: species.name,
      classification: species.classification,
      designation: species.designation,
      averageHeight: species.average_height,
      skinColors: species.skin_colors,
      hairColors: species.hair_colors,
      eyeColors: species.eye_colors,
      averageLifespan: species.average_lifespan,
      homeworld: species.homeworld,
      language: species.language,
      people: species.people,
      films: species.films,
      url: species.url,
    };
  }
}
