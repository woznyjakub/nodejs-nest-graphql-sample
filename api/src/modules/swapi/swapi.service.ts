import { HttpService } from '@nestjs/axios';
import { HttpException, Injectable } from '@nestjs/common';
import { ClassConstructor, plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { lastValueFrom } from 'rxjs';

import { FilmResponse, FilmsResponse } from './dto/film';
import { PeopleResponse } from './dto/people';
import { PlanetResponse, PlanetsResponse } from './dto/planet';
import { SpeciesManyResponse, SpeciesResponse } from './dto/species';
import { StarshipResponse, StarshipsResponse } from './dto/starship';
import { VehicleResponse, VehiclesResponse } from './dto/vehicle';

@Injectable()
export class SwapiService {
  constructor(private readonly httpService: HttpService) {}

  private readonly baseUrl = 'https://swapi.dev/api';

  private async getResource<T extends object>(
    name: string,
    dto: ClassConstructor<T>,
    id?: number,
    page?: number,
  ): Promise<T> {
    const url = `${this.baseUrl}/${name}${id ? `/${id}` : ''}`;
    const params = !id && page ? { page } : null;

    try {
      const response = await lastValueFrom(this.httpService.get(url, { params }));
      const resourceData: unknown = response.data;

      const resource = plainToInstance<T, unknown>(dto, resourceData, {
        enableImplicitConversion: true,
        excludeExtraneousValues: true,
      });

      const errors = await validate(resource);
      if (errors.length > 0) {
        throw new Error(
          `Swapi resource validation failed: ${errors.map((e) => e.toString()).join(', ')}`,
        );
      }

      return resource;
    } catch (error) {
      throw new HttpException(
        `Failed to get swapi resource (${name}). Error: ${error.message}`,
        error.status,
      );
    }
  }

  getVehicle(id: number): Promise<VehicleResponse> {
    return this.getResource('vehicles', VehicleResponse, id);
  }

  getVehicles(page: number = 1): Promise<VehiclesResponse> {
    return this.getResource('vehicles', VehiclesResponse, undefined, page);
  }

  getFilm(id: number): Promise<FilmResponse> {
    return this.getResource('films', FilmResponse, id);
  }

  getFilms(page: number = 1): Promise<FilmsResponse> {
    return this.getResource('films', FilmsResponse, undefined, page);
  }

  getStarship(id: number): Promise<StarshipResponse> {
    return this.getResource('starships', StarshipResponse, id);
  }

  getStarships(page: number = 1): Promise<StarshipsResponse> {
    return this.getResource('starships', StarshipsResponse, undefined, page);
  }

  getPlanet(id: number): Promise<PlanetResponse> {
    return this.getResource('planets', PlanetResponse, id);
  }

  getPlanets(page: number = 1): Promise<PlanetsResponse> {
    return this.getResource('planets', PlanetsResponse, undefined, page);
  }

  getSpecies(id: number): Promise<SpeciesResponse> {
    return this.getResource('species', SpeciesResponse, id);
  }

  getSpeciesMany(page: number = 1): Promise<SpeciesManyResponse> {
    return this.getResource('species', SpeciesManyResponse, undefined, page);
  }

  getPeople(page: number = 1): Promise<PeopleResponse> {
    return this.getResource('people', PeopleResponse, undefined, page);
  }
}
