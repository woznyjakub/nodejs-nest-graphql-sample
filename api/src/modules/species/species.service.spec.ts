import { Test, TestingModule } from '@nestjs/testing';

import { speciesResponseMock, speciesManyResponseMock } from '../../test/mocks/swapi';
import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SwapiService } from '../swapi/swapi.service';

import { SpeciesService } from './species.service';

describe('SpeciesService', () => {
  let speciesManyService: SpeciesService;

  const mockFetchWithCaching = vi.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: StarWarsCommonService,
          useValue: {
            fetchWithCaching: mockFetchWithCaching,
          },
        },
        SpeciesService,
        {
          provide: SwapiService,
          useValue: {
            getSpecies: vi.fn(),
            getSpeciesMany: vi.fn(),
          },
        },
      ],
    }).compile();

    speciesManyService = module.get<SpeciesService>(SpeciesService);
  });

  describe('single species data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(speciesResponseMock);

      const result = await speciesManyService.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple speciesMany data', () => {
    it('should return mapped data when', async () => {
      mockFetchWithCaching.mockReturnValue(speciesManyResponseMock);

      const result = await speciesManyService.findAll(1);

      expect(result).toMatchSnapshot();
    });
  });
});
