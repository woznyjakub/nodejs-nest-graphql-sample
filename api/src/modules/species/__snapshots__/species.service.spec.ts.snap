// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpeciesService > multiple speciesMany data > should return mapped data when 1`] = `
{
  "count": 37,
  "next": "https://swapi.dev/api/species/?page=2",
  "previous": null,
  "results": [
    {
      "averageHeight": "180",
      "averageLifespan": "120",
      "classification": "mammal",
      "designation": "sentient",
      "eyeColors": "brown, blue, green, hazel, grey, amber",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "hairColors": "blonde, brown, black, red",
      "homeworld": "https://swapi.dev/api/planets/9/",
      "language": "Galactic Basic",
      "name": "Human",
      "people": [
        "https://swapi.dev/api/people/66/",
        "https://swapi.dev/api/people/67/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/74/",
      ],
      "skinColors": "caucasian, black, asian, hispanic",
      "url": "https://swapi.dev/api/species/1/",
    },
    {
      "averageHeight": "n/a",
      "averageLifespan": "indefinite",
      "classification": "artificial",
      "designation": "sentient",
      "eyeColors": "n/a",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "hairColors": "n/a",
      "homeworld": null,
      "language": "n/a",
      "name": "Droid",
      "people": [
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/23/",
      ],
      "skinColors": "n/a",
      "url": "https://swapi.dev/api/species/2/",
    },
    {
      "averageHeight": "210",
      "averageLifespan": "400",
      "classification": "mammal",
      "designation": "sentient",
      "eyeColors": "blue, green, yellow, brown, golden, red",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/",
      ],
      "hairColors": "black, brown",
      "homeworld": "https://swapi.dev/api/planets/14/",
      "language": "Shyriiwook",
      "name": "Wookie",
      "people": [
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/80/",
      ],
      "skinColors": "gray",
      "url": "https://swapi.dev/api/species/3/",
    },
    {
      "averageHeight": "170",
      "averageLifespan": "unknown",
      "classification": "sentient",
      "designation": "reptilian",
      "eyeColors": "black",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "hairColors": "n/a",
      "homeworld": "https://swapi.dev/api/planets/23/",
      "language": "Galatic Basic",
      "name": "Rodian",
      "people": [
        "https://swapi.dev/api/people/15/",
      ],
      "skinColors": "green, blue",
      "url": "https://swapi.dev/api/species/4/",
    },
    {
      "averageHeight": "300",
      "averageLifespan": "1000",
      "classification": "gastropod",
      "designation": "sentient",
      "eyeColors": "yellow, red",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
      ],
      "hairColors": "n/a",
      "homeworld": "https://swapi.dev/api/planets/24/",
      "language": "Huttese",
      "name": "Hutt",
      "people": [
        "https://swapi.dev/api/people/16/",
      ],
      "skinColors": "green, brown, tan",
      "url": "https://swapi.dev/api/species/5/",
    },
    {
      "averageHeight": "66",
      "averageLifespan": "900",
      "classification": "mammal",
      "designation": "sentient",
      "eyeColors": "brown, green, yellow",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "hairColors": "brown, white",
      "homeworld": "https://swapi.dev/api/planets/28/",
      "language": "Galactic basic",
      "name": "Yoda's species",
      "people": [
        "https://swapi.dev/api/people/20/",
      ],
      "skinColors": "green, yellow",
      "url": "https://swapi.dev/api/species/6/",
    },
    {
      "averageHeight": "200",
      "averageLifespan": "unknown",
      "classification": "reptile",
      "designation": "sentient",
      "eyeColors": "yellow, orange",
      "films": [
        "https://swapi.dev/api/films/2/",
      ],
      "hairColors": "none",
      "homeworld": "https://swapi.dev/api/planets/29/",
      "language": "Dosh",
      "name": "Trandoshan",
      "people": [
        "https://swapi.dev/api/people/24/",
      ],
      "skinColors": "brown, green",
      "url": "https://swapi.dev/api/species/7/",
    },
    {
      "averageHeight": "160",
      "averageLifespan": "unknown",
      "classification": "amphibian",
      "designation": "sentient",
      "eyeColors": "yellow",
      "films": [
        "https://swapi.dev/api/films/3/",
      ],
      "hairColors": "none",
      "homeworld": "https://swapi.dev/api/planets/31/",
      "language": "Mon Calamarian",
      "name": "Mon Calamari",
      "people": [
        "https://swapi.dev/api/people/27/",
      ],
      "skinColors": "red, blue, brown, magenta",
      "url": "https://swapi.dev/api/species/8/",
    },
    {
      "averageHeight": "100",
      "averageLifespan": "unknown",
      "classification": "mammal",
      "designation": "sentient",
      "eyeColors": "orange, brown",
      "films": [
        "https://swapi.dev/api/films/3/",
      ],
      "hairColors": "white, brown, black",
      "homeworld": "https://swapi.dev/api/planets/7/",
      "language": "Ewokese",
      "name": "Ewok",
      "people": [
        "https://swapi.dev/api/people/30/",
      ],
      "skinColors": "brown",
      "url": "https://swapi.dev/api/species/9/",
    },
    {
      "averageHeight": "180",
      "averageLifespan": "unknown",
      "classification": "mammal",
      "designation": "sentient",
      "eyeColors": "black",
      "films": [
        "https://swapi.dev/api/films/3/",
      ],
      "hairColors": "none",
      "homeworld": "https://swapi.dev/api/planets/33/",
      "language": "Sullutese",
      "name": "Sullustan",
      "people": [
        "https://swapi.dev/api/people/31/",
      ],
      "skinColors": "pale",
      "url": "https://swapi.dev/api/species/10/",
    },
  ],
}
`;

exports[`SpeciesService > single species data > should return mapped data 1`] = `
{
  "averageHeight": "n/a",
  "averageLifespan": "indefinite",
  "classification": "artificial",
  "designation": "sentient",
  "eyeColors": "n/a",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/2/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/4/",
    "https://swapi.dev/api/films/5/",
    "https://swapi.dev/api/films/6/",
  ],
  "hairColors": "n/a",
  "homeworld": null,
  "language": "n/a",
  "name": "Droid",
  "people": [
    "https://swapi.dev/api/people/2/",
    "https://swapi.dev/api/people/3/",
    "https://swapi.dev/api/people/8/",
    "https://swapi.dev/api/people/23/",
  ],
  "skinColors": "n/a",
  "url": "https://swapi.dev/api/species/2/",
}
`;
